### DISCUSSION TITLE
portainer & portainer-agent swarm mode help

### DISCUSSION THREAD

**Discussion by ErgoProxy1337** [Tecnativa/docker-socket-proxy#57] [Q&A]
Hi there,

I am trying to set it up within the swarm but its a bit confusing on what I exactly need to change there or how to set it up.
Do u have any Informations about this?

```
  socket-proxy:
    container_name: socket-proxy
    image: tecnativa/docker-socket-proxy
    restart: always
    networks:
      socket_proxy:
        ipv4_address: xyz
    privileged: true
    ports:
      - "2375:2375"
    volumes:
      - "/var/run/docker.sock:/var/run/docker.sock:ro"
    environment:
      - LOG_LEVEL=debug # debug,info,notice,warning,err,crit,alert,emerg
      ## Variables match the URL prefix (i.e. AUTH blocks access to /auth/* parts of the API, etc.).
      # 0 to revoke access.
      # 1 to grant access.
      ## Granted by Default
      - EVENTS=1
      - PING=1
      - VERSION=1
      ## Revoked by Default
      # Security critical
      - AUTH=0
      - SECRETS=0
      - POST=0 # Ouroboros auf 1 falls benötigt
      # Not always needed
      - BUILD=0
      - COMMIT=0
      - CONFIGS=0
      - CONTAINERS=1 # Traefik, portainer, etc.
      - DISTRIBUTION=0
      - EXEC=0
      - IMAGES=1 # Portainer
      - INFO=1 # Portainer
      - NETWORKS=1 # Portainer
      - NODES=0
      - PLUGINS=0
      - SERVICES=1 # Portainer
      - SESSION=0
      - SWARM=1 # Needed auf 1?
      - SYSTEM=0
      - TASKS=1 # Portaienr
      - VOLUMES=1 # Portainer
    deploy:
      restart_policy:
        condition: on-failure
      placement:
        constraints: [node.role == manager]

  agent:
    image: portainer/agent
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock # command: -H tcp://socket-proxy:2375 <= needed?
      - /var/lib/docker/volumes:/var/lib/docker/volumes
    networks:
      - agent_network
    deploy:
      mode: global
      placement:
        constraints: [node.platform.os == linux]

  portainer:
    image: portainer/portainer-ce
    command: -H tcp://tasks.agent:9001 --tlsskipverify # // command: -H tcp://socket-proxy:2375 <= needed?
    ports:
      - "9000:9000"
      - "8000:8000"
    environment:
      - TZ=$TZ
      - DOCKER_HOST=socket-proxy:2375
    volumes:
      - portainer_data:/data
    networks:
      - agent_network
      - t2_proxy
      - socket_proxy
    deploy:
      mode: replicated
      replicas: 1
      placement:
        constraints: [node.role == manager]

volumes:
  portainer_data:
```

Are those changes correct and would the agent work on other nodes?

**Comment by tehniemer** [comment]
Did you get this figured out? I'm also trying to set up a portainer-agent with socket-proxy and am unable to get it connected to my main portainer instance. I'm able to get them connected if I don't use socket-proxy with the agent, but not if I do.

**Reply by ErgoProxy1337** [reply]
> Did you get this figured out? I'm also trying to set up a portainer-agent with socket-proxy and am unable to get it connected to my main portainer instance. I'm able to get them connected if I don't use socket-proxy with the agent, but not if I do.

Hey. sadly not rly...
I ended up in going back to the roots without swarm :(
I was able to make everything working but the additional complexity wasnt worth it for my purpose. Still interested in the solution if someone could help

**Reply by tehniemer** [reply]
Were you able to connect your portainer-agent to portainer while using socket-proxy with both? I'm interested in that bit more than swarm.

**Comment by grasuoare** [comment]
Up with this please. Also in the same case..

**Comment by maxfield-allison** [comment]
Finally did enough research on this to come to what I consider an answer. 

Portainer agent acts as a socket proxy itself so there is no need to add another socket proxy to this equation. 

That said, if you want to proxy anything other than portainer, you'll still need to run socket proxy and direct your other containers e.g. traefik2 to the socket proxy address. I'm just getting this built up in my cluster so I'll come back and update if I find anything else of use.

**Reply by snowy-jaguar** [reply]
Did you have a resolution to this?

I'm just wondering if there's anything special I need to consider when using docker-socket-proxy to proxy the swarm socket(s), I'm not going to try to use it to proxy the portainer socket as that seems redundant.
